### PR TITLE
Update grammar to accept user defined literal suffixes

### PR DESCRIPTION
**Pull request by matheusaaguiar**
Task of https://github.com/ethereum/solidity/pull/12656.

### CODE REVIEW

**Comment by matheusaaguiar** [comment]
AFAIU, whitespaces are ignored in the lexer and then values like the one below (from a failing test) are parsed by `antlr` with no errors:
https://github.com/ethereum/solidity/blob/ce63256c113fd88f185d65f76eaa45b683ecfb42/test/libsolidity/syntaxTests/literalSuffixes/application/invalid_suffix_no_whitespace.sol#L4

The same already happened before for `denominations`, but we don't seem to have any tests for this case. The following will also be parsed successfully by `antlr`:
```solidity
contract C {
 uint x = 1000gwei;
}
```

File: docs/grammar/SolidityParser.g4 [hunk 1]
```diff
@@ -409,10 +411,10 @@ inlineArrayExpression: LBrack (expression ( Comma expression)* ) RBrack;
 /**
  * Besides regular non-keyword Identifiers, some keywords like 'from' and 'error' can also be used as identifiers.
  */
-identifier: Identifier | From | Error | Revert | Global;
+identifier: Identifier | From | Error | Revert | Global | Suffix;
 
-literal: stringLiteral | numberLiteral | booleanLiteral | hexStringLiteral | unicodeStringLiteral;
-booleanLiteral: True | False;
+literal: (stringLiteral | numberLiteral | booleanLiteral | hexStringLiteral | unicodeStringLiteral) identifier?;
+booleanLiteral: (True | False);
```
Comment (cameel) [inline docs/grammar/SolidityParser.g4]: Does not look like the parentheses are really needed here.
```suggestion
booleanLiteral: True | False;
```

File: docs/grammar/SolidityParser.g4 [hunk 2]
```diff
@@ -409,10 +411,10 @@ inlineArrayExpression: LBrack (expression ( Comma expression)* ) RBrack;
 /**
  * Besides regular non-keyword Identifiers, some keywords like 'from' and 'error' can also be used as identifiers.
  */
-identifier: Identifier | From | Error | Revert | Global;
+identifier: Identifier | From | Error | Revert | Global | Suffix;
 
-literal: stringLiteral | numberLiteral | booleanLiteral | hexStringLiteral | unicodeStringLiteral;
-booleanLiteral: True | False;
+literal: (stringLiteral | numberLiteral | booleanLiteral | hexStringLiteral | unicodeStringLiteral) identifier?;
```
Comment (cameel) [inline docs/grammar/SolidityParser.g4]: You need to adjust the definition of `numberLiteral` or the grammar will now accept literals with both a denomination and a suffix specified, e.g. `1 gwei suffix;`.

I see two ways to do that:
1. Use a flag like `suffixSpecifierSet`.
1. Remove `NumberUnit` from `numberLiteral` and create a separate `numberLiteralWithDenomination` that includes it. Then have separate definitions for `literal` and `suffixedLiteral`, one including the denomination, the other not including it.

Comment (cameel) [reply docs/grammar/SolidityParser.g4]: Also, please add a syntax test covering this case (with both denomination and suffix, in various orders). I did not even think to try it :)

**Review by cameel — CHANGES_REQUESTED** [review]

**Comment by cameel** [comment]
> AFAIU, whitespaces are ignored in the lexer and then values like the one below (from a failing test) are parsed by `antlr` with no errors

I see we have a `WS` token that represents whitespace. Try putting it before identifier and see if that makes ANTLR require whitespace there.

Also, please add a syntax test for each of the whitespace characters that is allowed there - mine only has the space.

> The same already happened before for `denominations`, but we don't seem to have any tests for this case.

Good find. We should also do the same for subdenominations: fix the grammar and add syntax tests.

File: docs/grammar/SolidityParser.g4 (hunk 2 above)
Comment (cameel) [inline docs/grammar/SolidityParser.g4]: While at it, we should also rename `NumberUnit` to `Subdenomination`. We use "subdenomination" in the code and the grammar seems to be the only place where a "number unit" is used. In the docs we don't have a common name and refer to some of them as [ether units](https://docs.soliditylang.org/en/latest/units-and-global-variables.html#ether-units) and others as [time units](https://docs.soliditylang.org/en/latest/units-and-global-variables.html#time-units).

Comment (cameel) [reply docs/grammar/SolidityParser.g4]: Actually, @chriseth anything against this renaming?

Comment (chriseth) [reply docs/grammar/SolidityParser.g4]: sounds good

**Comment by matheusaaguiar** [comment]
> I see we have a WS token that represents whitespace. Try putting it before identifier and see if that makes ANTLR require whitespace there.

I tried it but doesn't work.
In the previous message, I should have explained better that the `WS` token defined in the lexer recognizes whitespaces and then discards them using the `skip` [command](https://github.com/antlr/antlr4/blob/master/doc/lexer-rules.md#skip). So, when the parser gets the tokens, none of them are `WS` and thus it never chooses `suffixedLiteral`  or `numberLiteralWithSubdenomination` rules.

I also found a problem with the `functionDefinition` rule. Since it can recognize multiple `modifierInvocation` and `suffix` is not a language keyword, the `antlr` parser will fail to error out a function with repeated `suffix` specifiers:
```solidity
function suffix1(uint) pure suffix suffix returns (uint) {}
// ----
// ParserError 2878: (82-88): Suffix already specified.
```

**Comment by cameel** [comment]
> I also found a problem with the `functionDefinition` rule. Since it can recognize multiple `modifierInvocation` and `suffix` is not a language keyword, the `antlr` parser will fail to error out a function with repeated `suffix` specifiers:

Hmm... Indeed looks like a problem. Actually, why `virtual` and `override` do not cause similar problems? The current grammar allows them on free function definitions. Do we just have no syntax test that tries to put `virtual` on a free function?

But anyway, one way out would be to split `functionDefinition` into `freeFunctionDefinition` and `contractFunctionDefinition`. It actually would not be that bad because it would make the grammar more precise and `freeFunctionDefinition` would be quite simple because it does not allow a lot of stuff that's allowed in contract functions.

Another one would be to allow only either `modifierInvocation` or `Suffix`. But I suspect that the only nice way to achieve that in ANTLR may be the split I mentioned above.

**Comment by cameel** [comment]
> I tried it but doesn't work.

ok, so here's another idea. Looks like our scanner actually properly parses `1000suffix` as a number followed by a literal and just has an extra validation that rejects it: https://github.com/ethereum/solidity/blob/2ca349c69aefe8c7033764f2e49c91fe99ccaed5/liblangutil/Scanner.cpp#L1004-L1009

We could simulate this by recognizing a number followed by literal as its own thing. Try to define `DecimalNumberFollowedByIdentifier`, which can basically consists of chars allowed in `DecimalNumber`, and then chars allowed in identifier. Just defining such a rule should make ANTLR recognize that `1000suffix` is not a decimal number.

Also, to make sure it works, please add test cases for hex and fractional numbers (I suspect you'll need separate rules to handle them): `0x1000suffix`, `1000.0suffix` `1000.0e-5suffix` and `0x1000abcdefgh`.

**Comment by cameel** [comment]
Also, these grammar changes (i.e. split into free/contract function, number+identifier, renaming of `NumberUnit`) are quite general and do not depend on suffixes. Once you get them to pass all tests, I'd suggest to move them (along with relevant test cases) to another PR, straight to `develop` and here leave only changes that introduce `suffix`. We can get them merged separately and keep the main suffix PR smaller that way.

File: docs/grammar/SolidityParser.g4 [hunk 3]
```diff
@@ -409,9 +412,12 @@ inlineArrayExpression: LBrack (expression ( Comma expression)* ) RBrack;
 /**
  * Besides regular non-keyword Identifiers, some keywords like 'from' and 'error' can also be used as identifiers.
  */
-identifier: Identifier | From | Error | Revert | Global;
+identifier: Identifier | From | Error | Revert | Global | Suffix;
+
+suffixedLiteral: (stringLiteral | numberLiteral | booleanLiteral | hexStringLiteral | unicodeStringLiteral) identifier;
+
+literal: stringLiteral | numberLiteral | booleanLiteral | hexStringLiteral | unicodeStringLiteral | numberLiteralWithSubdenomination;
```
Comment (cameel) [inline docs/grammar/SolidityParser.g4]: Actually, we could simplify this and avoid having to introduce the extra rule:
```suggestion
literal: stringLiteral | numberLiteral Subdenomination? | booleanLiteral | hexStringLiteral | unicodeStringLiteral;
```

Comment (cameel) [inline docs/grammar/SolidityParser.g4]: We now have a bit of duplication here. Maybe something like this would be better?

```suggestion
literal: stringLiteral | numberLiteral | booleanLiteral | hexStringLiteral | unicodeStringLiteral;

suffixedLiteral: literal identifier;
literalWithSubdenomination: numberLiteral Subdenomination;
```

Then we'd explicitly list all variants of a literal in `expression` below:
```
		| literal
		| literalWithSubdenomination
		| suffixedLiteral
```

This might also be more readable because, personally, I'd not expect something like `1 gwei` to be parsed as a literal. I'd expect a separate rule for it.

File: test/libsolidity/syntaxTests/literalSuffixes/application/invalid_suffix_denomination.sol [hunk 4]
```diff
@@ -0,0 +1,7 @@
+function suffix(uint) pure suffix returns (uint) { return 1; }
+
+contract C {
+    uint x = 8 suffix gwei;
```
Comment (cameel) [inline test/libsolidity/syntaxTests/literalSuffixes/application/invalid_suffix_denomination.sol:4]: Please add also something like `8 gwei ether`.

**Comment by matheusaaguiar** [comment]
> Actually, why virtual and override do not cause similar problems? The current grammar allows them on free function definitions. Do we just have no syntax test that tries to put virtual on a free function?

We do have tests for that, but they generate `SyntaxError` rather than `ParserError`, which is the only category `test_antlr_grammar.sh` looks for in order to detect divergence with `solc` parser.
https://github.com/ethereum/solidity/blob/2ca349c69aefe8c7033764f2e49c91fe99ccaed5/scripts/test_antlr_grammar.sh#L83-L88


> But anyway, one way out would be to split functionDefinition into freeFunctionDefinition and contractFunctionDefinition.

Ok, agreed.

**Comment by cameel** [comment]
> We do have tests for that, but they generate `SyntaxError` rather than `ParserError`, which is the only category `test_antlr_grammar.sh` looks for in order to detect divergence with `solc` parser.

Interesting. Well, then maybe working around it `test_antlr_grammar.sh` in some way would not be out of the question if adjusting it turns out to be too complicated. But first let's try to do it the nice way. We should be able to match the grammar.

File: docs/grammar/SolidityParser.g4 (hunk 3 above)
Comment (matheusaaguiar) [reply docs/grammar/SolidityParser.g4]: Yes, that's much better and easier to read. Done.

File: test/libsolidity/syntaxTests/invalid_octal_number.sol
Comment (cameel) [inline test/libsolidity/syntaxTests/invalid_octal_number.sol:1]: This test should go into the `literals/` dir.

File: test/libsolidity/syntaxTests/invalid_hex_number.sol
Comment (cameel) [inline test/libsolidity/syntaxTests/invalid_hex_number.sol:1]: This test should go into the `literals/` dir.

File: docs/grammar/SolidityParser.g4 [hunk 5]
```diff
@@ -175,6 +175,25 @@ locals[
...
 	(Returns LParen returnParameters=parameterList RParen)?
 	(Semicolon | body=block);
+
+/** The definition of a free function.
+ *
+ */
+ freeFunctionDefinition
+ locals[
+ 	boolean mutabilitySet = false,
+ 	boolean suffixSpecifierSet = false
```
Comment (cameel) [inline docs/grammar/SolidityParser.g4]:
```suggestion
 	boolean suffixSpecifierSet = false,
```

**Review by cameel — COMMENTED** [review]
ok, looks good now apart from some minor tweaks.

Please now extract the generic part into a PR on develop as I suggested in an earlier comment (note that this will require changing some tests to use denominations rather than suffixes).

**Comment by cameel** [comment]
By the way, please remember to mark PRs as reviewable once you're done implementing them.

**Comment by cameel** [comment]
I rebased the suffix PR on #14066, so if you rebase this one on the suffix PR, we can now continue review here.

File: docs/grammar/SolidityParser.g4 [hunk 6]
```diff
@@ -179,10 +179,18 @@ locals[
...
+ 	boolean suffixSpecifierSet = false,
+ ]
+ :
  	Function (identifier | Fallback | Receive)
  	LParen (arguments=parameterList)? RParen
- 	stateMutability?
+ 	(
+		{!$mutabilitySet}? stateMutability {$mutabilitySet = true;}
+		| {!$suffixSpecifierSet}? Suffix {$suffixSpecifierSet = true;}
+ 	 )*
```
Comment (cameel) [inline docs/grammar/SolidityParser.g4]: With just mutability and suffix we could still do it without variables and keep it simple:
```suggestion
	stateMutability Suffix?
	| Suffix stateMutability?
```

File: docs/grammar/SolidityParser.g4 [hunk 7]
```diff
@@ -421,12 +430,14 @@ inlineArrayExpression: LBrack (expression ( Comma expression)* ) RBrack;
...
  * Besides regular non-keyword Identifiers, some keywords like 'from' and 'error' can also be used as identifiers.
  */
-identifier: Identifier | From | Error | Revert | Global;
+identifier: Identifier | From | Error | Revert | Global | Suffix;
 
 literal: stringLiteral | numberLiteral | booleanLiteral | hexStringLiteral | unicodeStringLiteral;
 
 literalWithSubDenomination: numberLiteral SubDenomination;
 
+suffixedLiteral: literal identifier;
```
Comment (cameel) [inline docs/grammar/SolidityParser.g4]: Actually, just realized that this is not entirely correct. It does not have to be a single literal. Can be member access as well.

**Review by cameel — CHANGES_REQUESTED** [review]

File: docs/grammar/SolidityParser.g4 (hunk 7 above)
Comment (matheusaaguiar) [reply docs/grammar/SolidityParser.g4]: Not sure if I follow here. Should we be doing something like this? 

```suggestion
memberAcess: identifier (Period identifier)+;

suffixedLiteral: (literal | memberAcess) identifier;
```

What would be a member access that is equivalent to a literal?

Comment (cameel) [reply docs/grammar/SolidityParser.g4]: I mean that something like `123 M.suffix` is allowed because you can do `import "A.sol" as M` and `A.sol` can have `suffix` defined in it. `identifier` alone does not allow it.

Not sure if we want to define `memberAccess` explicitly though. I see that inside `expression` that's defined in place like this:
```
expression Period (identifier | Address) # MemberAccess
```

So I guess the solution is either to repeat that for `suffixedLiteral` or indeed define it as `memberAccess`.

Comment (matheusaaguiar) [reply docs/grammar/SolidityParser.g4]: Ah, ok, understood it now. I was thinking about the `literal`, not the `identifier`.
We have a test with suffix function imported from other file:
https://github.com/ethereum/solidity/blob/65188cafc3a277c0dfa74fb09c17e16f99f19d31/test/libsolidity/syntaxTests/literalSuffixes/usableAsSuffix/imported_function_as_suffix.sol#L6-L14

I have removed the failing test in `chk_antlr_grammar` before running it locally and then got no errors, so I think that the current grammar already covers that somehow, but not sure.

File: docs/grammar/SolidityLexer.g4 [hunk 8]
```diff
@@ -85,11 +85,13 @@ SignedIntegerType:
...
 String: 'string';
 Struct: 'struct';
+Suffix: 'suffix';
 True: 'true';
 Try: 'try';
 Type: 'type';
 Ufixed: 'ufixed' | ('ufixed' [1-9][0-9]+ 'x' [1-9][0-9]+);
 Unchecked: 'unchecked';
+Unicode: 'unicode';
 /**
```
Comment (cameel) [inline docs/grammar/SolidityLexer.g4]: Oh, so a keyword was missing here. This is a general change that should go straight to develop.

With a test case preferably.

File: docs/grammar/SolidityParser.g4 (hunk 7 above)
Comment (cameel) [reply docs/grammar/SolidityParser.g4]: > I think that the current grammar already covers that somehow, but not sure.

But how? It can't be. `expression` does account for that explicitly so the literals should have to as well.

File: docs/grammar/SolidityLexer.g4 (hunk 8 above)
Comment (matheusaaguiar) [reply docs/grammar/SolidityLexer.g4]: Ok. Created #14078 .

**Comment by matheusaaguiar** [comment]
> But how? It can't be. expression does account for that explicitly so the literals should have to as well.

Yeah, I am not sure why it works. I will investigate it.

**Comment by cameel** [comment]
> Yeah, I am not sure why it works. I will investigate it.

So, did you figure it out?

**Comment by matheusaaguiar** [comment]
> So, did you figure it out?

Not yet. I will prioritize it though.

**Comment by matheusaaguiar** [comment]
I checked the `test_antlr_grammar` script and found out that tests that are multi-source files like the one I mentioned before are excluded before actually running the `ANTLR` parser.
https://github.com/ethereum/solidity/blob/65188cafc3a277c0dfa74fb09c17e16f99f19d31/test/libsolidity/syntaxTests/literalSuffixes/usableAsSuffix/imported_function_as_suffix.sol#L1-L5

Still, seems like the parser accepts member access as literal suffixes...

**Comment by cameel** [comment]
> Still, seems like the parser accepts member access as literal suffixes...

You mean ANTLR or solc parser? Because the latter obviously does, but ANTLR should too.

In any case, this explains why it did not fail. This means that the grammar is really incomplete, we just can't properly test it. It needs to be updated to cover this case.

**Comment by matheusaaguiar** [comment]
> You mean ANTLR or solc parser? 

I meant `ANTLR` parser. I just removed the `==== Source ...  ` separators from the file and then let the script apply the `ANTLR` parser on it, which reported success.

> It needs to be updated to cover this case.

Are we going to do this in the context of the literal suffix PR or is it a task we leave for later?

**Comment by cameel** [comment]
> Are we going to do this in the context of the literal suffix PR or is it a task we leave for later?

In this PR. Without it the grammar is incomplete.

> I just removed the `==== Source ... ` separators from the file and then let the script apply the `ANTLR` parser on it, which reported success.

This sounds weird. Is it only for suffixes or does it let you use dotted paths in other places that grammar does not officially include?

**Comment by matheusaaguiar** [comment]
I checked with the `ANTLR` test tool using the options `-tokens` and `-gui` (manually typing in CLI). When parsing expression `uint x = 1 A.b.c;` it produces the following output:

```
[@0,0:3='uint',<UnsignedIntegerType>,1:0]
[@1,5:12='constant',<'constant'>,1:5]
[@2,14:14='x',<Identifier>,1:14]
[@3,16:16='=',<'='>,1:16]
[@4,18:18='1',<DecimalNumber>,1:18]
[@5,20:20='A',<Identifier>,1:20]
[@6,21:21='.',<Period>,1:21]
[@7,22:22='b',<Identifier>,1:22]
[@8,23:23='.',<Period>,1:23]
[@9,24:24='c',<Identifier>,1:24]
[@10,25:25=';',<Semicolon>,1:25]
[@11,27:26='<EOF>',<EOF>,2:0]

``` 

The visualization tool produces this tree:
antlr4_parse_tree


It first recognizes the value of the assignment as a member access, and then decomposes the `expression` before the `.` as either another member access or finally as a `suffixedLiteral`.

**Comment by matheusaaguiar** [comment]
> This sounds weird. Is it only for suffixes or does it let you use dotted paths in other places that grammar does not officially include?

I am trying to find if there's other cases.

**Comment by cameel** [comment]
Oh, that's very helpful. I see what's happening there.

Basically, ANTLR sees it as if it was `uint x = (1 A).b.c;`, which is already a valid expression. solc parser on the other hand interprets it as `uint x = 1 (A.b.c);`. We do need an extra rule to get that second interpretation.

**Comment by cameel** [comment]
I pushed a new version of the base PR. Please rebase.

File: scripts/test_antlr_grammar.sh [hunk 9]
```diff
@@ -129,7 +129,8 @@ done < <(
       grep -v -E 'license/license_hidden_unicode.sol' |
       grep -v -E 'license/license_unicode.sol' |
       # Skipping tests with 'something.address' as 'address' as the grammar fails on those
-      grep -v -E 'inlineAssembly/external_function_pointer_address.*.sol'
+      grep -v -E 'inlineAssembly/external_function_pointer_address.*.sol' |
+      grep -v -E 'literalSuffixes/application/invalid_address_member_on_suffix.sol'
```
Comment (matheusaaguiar) [inline scripts/test_antlr_grammar.sh:133]: I added this because I noticed we already had a workaround for the case.

**Review by cameel — APPROVED** [review]
ok, looks good now!
